common/src/pause.rs

```diff
@@ -0,0 +1,125 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+use frame_support::storage::PrefixIterator;
+
+#[derive(Clone, Debug, Decode, Encode, PartialEq, TypeInfo)]
+struct PausedProgram {
+    program_id: H256,
+    program: ActiveProgram,
+    pages_hash: H256,
+    wait_list: Vec<QueuedDispatch>,
+    waiting_init: Vec<H256>,
+}
+
+fn decode_dispatch_tuple(_: &[u8], value: &[u8]) -> Result<(QueuedDispatch, u32), codec::Error> {
+    <(QueuedDispatch, u32)>::decode(&mut &*value)
+}
+
+fn memory_pages_hash(pages: &BTreeMap<u32, Vec<u8>>) -> H256 {
+    pages.using_encoded(sp_io::hashing::blake2_256).into()
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PauseError {
+    ProgramNotFound,
+    ProgramTerminated,
+}
+
+pub fn pause_program(program_id: H256) -> Result<(), PauseError> {
+    let program = get_program(program_id).ok_or(PauseError::ProgramNotFound)?;
+    let program: ActiveProgram = program
+        .try_into()
+        .map_err(|_| PauseError::ProgramTerminated)?;
+
+    let prefix = wait_prefix(program_id);
+    let previous_key = prefix.clone();
+
+    let paused_program = PausedProgram {
+        program_id,
+        pages_hash: memory_pages_hash(
+            &get_program_pages(program_id, program.persistent_pages.clone())
+                .expect("pause_program: active program exists, therefore pages do"),
+        ),
+        program,
+        wait_list: PrefixIterator::<_, ()>::new(prefix, previous_key, decode_dispatch_tuple)
```

Comment:
Why not use `frame_support::storage::KeyPrefixIterator` if you only need the key and discard the value? We have a similar example in the `usage` pallet (https://github.com/gear-tech/gear/blob/master/pallets/usage/src/offchain.rs#L124)

Reply:
If I'm not mistaken the situation is the opposite - only value is used.